wp-includes/mysql/class-mysql-lexer.php

```diff
@@ -0,0 +1,2752 @@
+		}
+	}
+
+	protected function hex_number() {
+		$is_quoted = 'x' === strtolower( $this->c ) && "'" === $this->n;
+
+		// Consume "0x" or "x'".
+		$this->consume();
+		$this->consume();
+
+		while (
+			( $this->c >= '0' && $this->c <= '9' )
+			|| ( $this->c >= 'a' && $this->c <= 'f' )
+			|| ( $this->c >= 'A' && $this->c <= 'F' )
+		) {
+			$this->consume();
+		}
+
+		if ( $is_quoted ) {
+			$this->consume(); // Consume the "'".
+		}
+
+		$this->type = self::HEX_NUMBER;
+	}
+
+	protected function bin_number() {
+		$is_quoted = 'b' === strtolower( $this->c ) && "'" === $this->n;
+
+		// Consume "0b" or "b'".
+		$this->consume();
+		$this->consume();
+
+		while ( '0' === $this->c || '1' === $this->c ) {
+			$this->consume();
+		}
+
+		if ( $is_quoted ) {
+			$this->consume(); // Consume the "'".
+		}
+
+		$this->type = self::BIN_NUMBER;
+	}
+
+	protected function int_number() {
+		while ( $this->is_digit( $this->c ) ) {
+			$this->consume();
+		}
+		$this->type = self::INT_NUMBER;
+	}
+
+	protected function decimal_number() {
+		$this->consume(); // Consume the '.'.
+		while ( $this->is_digit( $this->c ) ) {
+			$this->consume();
+		}
+		$this->type = self::DECIMAL_NUMBER;
+	}
+
+	protected function line_comment() {
+		while ( null !== $this->c && "\n" !== $this->c && "\r" !== $this->c ) {
+			$this->consume();
+		}
+		$this->type    = self::COMMENT;
+		$this->channel = self::CHANNEL_HIDDEN;
+	}
+
+	protected function block_comment() {
+		$this->consume(); // Consume the '/'.
+		$this->consume(); // Consume the '*'.
+		$this->skip_comment_content();
+		$this->type    = self::COMMENT;
+		$this->channel = self::CHANNEL_HIDDEN;
+	}
+
+	protected function mysql_comment() {
+		// MySQL-specific comment in one of the following forms:
+		//   1. /*! ... */      - The content is treated as regular SQL code.
+		//   2. /*!12345 ... */ - The content is treated as SQL code when "server version >= 12345".
+		$this->consume(); // Consume the '/'.
+		$this->consume(); // Consume the '*'.
+		$this->consume(); // Consume the '!'.
+
+		// Check if the next 5 characters are digits.
+		$is_version_comment = true;
+		for ( $i = 0; $i < 5; $i++ ) {
+			if ( ! $this->is_digit( $this->c ) ) {
+				$is_version_comment = false;
+				break;
+			}
+			$this->consume(); // Consume the digit.
+		}
+
+		// For version comments, extract the version number.
+		$version = $is_version_comment
+			? (int) substr( $this->get_text(), 3 ) // Strip the '/*!' prefix.
+			: 0;
+
+		if ( $this->server_version < $version ) {
+			// When version not satisfied. Treat the content as a regular comment.
+			$this->skip_comment_content();
+			$this->type = self::COMMENT;
+		} else {
+			// Version satisfied or not specified. Treat the content as SQL code.
+			$this->in_version_comment = true;
+			$this->type               = self::MYSQL_COMMENT_START;
+		}
+		$this->channel = self::CHANNEL_HIDDEN;
+	}
+
+	protected function skip_comment_content() {
+		while ( null !== $this->c ) {
```

Comment:
This could be replaced with `strcspn($this->input, '*', $this->position)` and one more check if the character after that is `/`.

Reply:
Also, it took me a moment to figure out what's `null !== $this->c`. How about `$this->position < strlen( $this->sql )`? AFAIR `strlen()` is really fast and just returns a precomputed length of the string.